### ISSUE TITLE
The use of validation dataloader changes the train dataloader 

### ISSUE THREAD

**Issue by zanshuxun** [pytorch/pytorch#80119]
### 🐛 Describe the bug

The use of validation dataloader changes the train dataloader, which is unexpected.

## Without validation

``` python
import torch
import torch.utils.data as data
from torch.utils.data import DataLoader

torch.manual_seed(2022)

train_dataset = data.TensorDataset(torch.randn(5, 1))
val_dataset = data.TensorDataset(torch.randn(2, 1))

train_loader = DataLoader(dataset=train_dataset, shuffle=True, batch_size=1)
val_loader = DataLoader(dataset=val_dataset, shuffle=False, batch_size=1)

for epoch in range(3):
    print('\nepoch', epoch)    
    # train
    for _, x in enumerate(train_loader):
        print(x[0])

    # val
    # for _, x in enumerate(val_loader):
    #     pass
```

output: (the result is **reproducible** since I set the `torch.manual_seed(2022)`)
```
epoch 0
tensor([[0.8936]])
tensor([[-0.2044]])
tensor([[0.1915]])
tensor([[0.3306]])
tensor([[0.2306]])

epoch 1
tensor([[0.8936]])
tensor([[0.1915]])
tensor([[0.3306]])
tensor([[-0.2044]])
tensor([[0.2306]])

epoch 2
tensor([[0.3306]])
tensor([[0.2306]])
tensor([[-0.2044]])
tensor([[0.1915]])
tensor([[0.8936]])
```

## With validation

When I use the validation dataloader after each epoch, the following training process changed:

``` python
import torch
import torch.utils.data as data
from torch.utils.data import DataLoader

torch.manual_seed(2022)

train_dataset = data.TensorDataset(torch.randn(5, 1))
val_dataset = data.TensorDataset(torch.randn(2, 1))

train_loader = DataLoader(dataset=train_dataset, shuffle=True, batch_size=1)
val_loader = DataLoader(dataset=val_dataset, shuffle=False, batch_size=1)

for epoch in range(3):
    print('\nepoch', epoch)    
    # train
    for _, x in enumerate(train_loader):
        print(x[0])

    # val
    for _, x in enumerate(val_loader):
        pass
```

output: (the data order has been **changed** in epoch 1 and epoch 2)
```
epoch 0
tensor([[0.8936]])
tensor([[-0.2044]])
tensor([[0.1915]])
tensor([[0.3306]])
tensor([[0.2306]])

epoch 1
tensor([[-0.2044]])
tensor([[0.1915]])
tensor([[0.2306]])
tensor([[0.3306]])
tensor([[0.8936]])

epoch 2
tensor([[-0.2044]])
tensor([[0.8936]])
tensor([[0.3306]])
tensor([[0.2306]])
tensor([[0.1915]])
```

I think the validation process should not have influence on training process, and hope to solve this problem.

### Versions

```
Collecting environment information...
PyTorch version: 1.10.2
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.4 (x86_64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.9)
CMake version: Could not collect
Libc version: N/A

Python version: 3.6.5 |Anaconda, Inc.| (default, Apr 26 2018, 08:42:37)  [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] (64-bit runtime)
Python platform: Darwin-20.5.0-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] deepctr-torch==0.2.8
[pip3] numpy==1.19.2
[pip3] numpydoc==0.8.0
[pip3] torch==1.10.2
[conda] blas                      1.0                         mkl  
[conda] deepctr-torch             0.2.8                     <pip>
[conda] mkl                       2018.0.2                      1  
[conda] mkl-service               1.1.2            py36h7ea6df4_4  
[conda] mkl_fft                   1.0.1            py36h917ab60_0  
[conda] mkl_random                1.0.1            py36h78cc56f_0  
[conda] numpy                     1.14.3           py36h9bb19eb_1  
[conda] numpy                     1.19.2                    <pip>
[conda] numpy-base                1.14.3           py36h479e554_1  
[conda] numpydoc                  0.8.0                    py36_0  
[conda] torch                     1.10.2                    <pip>
```

cc @SsnL @VitalyFedyunin @ejguan @NivekT

**Comment by NivekT** [comment]
You can pass in a `torch.Generator` that is not used by other objects.
```python
generator = torch.Generator()
generator.manual_seed(2022)
train_loader = DataLoader(dataset=train_dataset, shuffle=True, batch_size=1, generator=generator)
# Don't pass in the same generator to `valid_loader` if you don't want the RNG to share states and influence one another.
```

cc: @ejguan

**Comment by ejguan** [comment]
The reason simply because a random `base_seed` is lazily generated when iteration starts even though you set `shuffle=False`.

Like @NivekT said, you should provide a generator with a manual seed to `train_dataloader` for randomness control. Besides, if you have random operations in your model after loading data, the randomness of `train_dataloader` is also affected. Providing generator to isolate random operation from other components would be the best option for you.

**Comment by NivekT** [comment]
I am closing this issue for now.  Please feel free to re-open if you have further questions or suggestions for improvements. Thanks!

**Comment by zanshuxun** [comment]
@NivekT @ejguan Thank you for your help

**Comment by zanshuxun** [comment]
@NivekT  Hi Kevin, in In some lower versions of PyTorch (e.g.,1.5.0), there is no `generator` parameter in `DataLoader`, then what should I do?

![image](https://user-images.githubusercontent.com/28911886/175805243-31c79a87-a0cc-4ed0-bf45-173314e49e8a.png)

![image](https://user-images.githubusercontent.com/28911886/175805949-de5c73c6-d7bf-40c2-b244-f7baaaf5f266.png)

Versions:
```
Collecting environment information...
PyTorch version: 1.5.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 企业版
GCC version: (tdm64-1) 4.9.2
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.6.5 |Anaconda, Inc.| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.18362-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] numpy==1.19.3
[pip3] numpydoc==0.8.0
[pip3] torch==1.5.0+cpu
[conda] blas                      1.0                         mkl    defaults
[conda] mkl                       2018.0.2                      1    defaults
[conda] mkl-service               1.1.2            py36h57e144c_4    defaults
[conda] mkl_fft                   1.0.1            py36h452e1ab_0    defaults
[conda] mkl_random                1.0.1            py36h9258bd6_0    defaults
[conda] numpy                     1.19.3                    <pip>
[conda] numpy                     1.14.3           py36h9fa60d3_1    defaults
[conda] numpy-base                1.14.3           py36h555522e_1    defaults
[conda] numpydoc                  0.8.0                    py36_0    defaults
[conda] torch                     1.5.0+cpu                 <pip>

Process finished with exit code 0

```
